snuba/query/expressions.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Callable, Iterator, Optional, Sequence
+
+from snuba.query.collections import NodeContainer
+
+
+@dataclass
+class Expression(ABC):
+    """
+    A node in the Query AST. This can be a leaf or an intermediate node.
+    It represents an expression that can be resolved to a value. This
+    includes column names, function calls and boolean conditions (which are
+    function calls themselves in the AST), literals, etc.
+
+    The root of the tree is not a Node itself yet (since it is the Query object).
+    Representing the root as a node itself does not seem very useful right now
+    since we never traverse the full tree. We could revisit that later.
+
+    All expression can have an optional alias.
+    """
+
+    alias: Optional[str]
+
+    @abstractmethod
+    def format(self) -> str:
+        """
+        Turn this node into a string for the Clickhouse query.
+
+        TODO: provide a clickhouse formatter to this method so that, through
+        a strategy pattern, this class will provide the content for the query
+        and the clickhouse formatter will provide the format.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        """
+        Transforms this expression through the function passed in input.
+        This works almost like a map function over sequences though, contrarily to
+        sequences, this acts on a subtree. The semantics of transform can be different
+        between intermediate nodes and leaves, so each node class can implement it
+        its own way.
+
+        It returns the new expression. It does not guarantee the returned value
+        will be a new object. No guarantee of immutability is given on expressions.
```

Comment:
Let me break the comment into two:
1) "It does not guarantee the returned value will be a new object." This is inevitable since it depends on what the function passed does. On a leaf node, if `func` returns self, it is the same object. If it builds a different object instead of mutating the existing then it is a new object.
2) "No guarantee of immutability is given on expressions." If we do not impose immutability, the point at number 1 implies that a transform may return the current object mutated or a new one, we cannot do anything about that. We can try to force all expressions to be immutable. This would be trivial for leaves, but more complex for expresisons that have children since children are a sequence. In order to make HierarchicalExpression immutable, we need to be able to tell either that the transformation of the children did not produce a change, thus we can return self (feasible) or always assume that we would instantiate a separate HierarchicalExpression every time transform is called on it, whether the content was changed or not, which does not seem very efficient. 

Does it make sense ?